feat: add no compatibility check option for enabling experiments

SLASHLogin opened the pull request:
This would skip check for Ubuntu distro and experiment specific checks.

This would satisfy needs from the #7 and #11 to run this on distros which are not supported.

jnsgruk commented:
Thanks for doing this!

I think I'd prefer to just add `no_compatibility_check` to the `enable` and `disable` methods on the `enum` implementation - and perhaps we can skip the confirmation - something like this:

```diff
diff --git i/src/experiments/mod.rs w/src/experiments/mod.rs
index 02e5b72..6326563 100644
--- i/src/experiments/mod.rs
+++ w/src/experiments/mod.rs
@@ -20,8 +20,8 @@ impl Experiment<'_> {
         }
     }
 
-    pub fn enable(&self) -> Result<()> {
-        if !self.check_compatible() {
+    pub fn enable(&self, no_compatibility_check: bool) -> Result<()> {
+        if no_compatibility_check || !self.check_compatible() {
             warn!(
                 "Skipping '{}'. Minimum supported release is {}.",
                 self.name(),
diff --git i/src/main.rs w/src/main.rs
index f1d1e2f..e792630 100644
--- i/src/main.rs
+++ w/src/main.rs
@@ -67,6 +67,14 @@ struct Args {
     )]
     all: bool,
 
+    #[arg(
+        long,
+        default_value_t = false,
+        global = true,
+        help = "Skip compatibility checks (dangerous)"
+    )]
+    no_compatibility_check: bool,
+
     #[arg(
         short,
         long,
@@ -119,29 +127,38 @@ fn main() -> Result<()> {
 
     // Handle subcommands
     match args.cmd {
-        Commands::Enable => enable(&system, selected, args.yes),
-        Commands::Disable => disable(selected, args.yes),
+        Commands::Enable => enable(&system, selected, args.yes, args.no_compatibility_check),
+        Commands::Disable => disable(selected, args.yes, args.no_compatibility_check),
     }
 }
 
 /// Enables selected experiments
-fn enable(system: &impl Worker, experiments: Vec<Experiment>, yes: bool) -> Result<()> {
+fn enable(
+    system: &impl Worker,
+    experiments: Vec<Experiment>,
+    yes: bool,
+    no_compatibility_check: bool,
+) -> Result<()> {
     confirm_or_exit(yes);
 
     info!("Updating apt package cache");
     system.update_package_lists()?;
 
     for e in experiments.iter() {
-        e.enable()?;
+        e.enable(no_compatibility_check)?;
     }
     Ok(())
 }
 
 // Disable selected experiments
-fn disable(experiments: Vec<Experiment<'_>>, yes: bool) -> Result<()> {
+fn disable(
+    experiments: Vec<Experiment<'_>>,
+    yes: bool,
+    no_compatibility_check: bool,
+) -> Result<()> {
     confirm_or_exit(yes);
     for e in experiments.iter() {
-        e.disable()?;
+        e.disable(no_compatibility_check)?;
     }
     Ok(())
 }
```

Before this can merge, it would also be useful to add to the test suite.

SLASHLogin commented:
Removed extra confirmation and added one test. I can see that it would be run on Ubuntu, would you like to have new workflow file/suite for other distros, or this should do?

jnsgruk commented:
Okay this is looking good! Just need to handle the failure on 24.04 in the test, since I'd expect it to fail on account of the package not being present 😃

SLASHLogin commented:
Oh yeah, just realized that we've disabled the compatibility check and tried to install the package where it's not available :D

jnsgruk reviewed (approved):
Looks good, thank you!

SLASHLogin commented:
Thanks a lot for approving this and creating this tool! Have a wonderful day!
